sno/dataset2_gpkg.py

```diff
@@ -1,17 +1,90 @@
 import re
 from .dataset2 import Schema, ColumnSchema
 
-_GPKG_TYPE_TO_V2_TYPE = {
-    "SMALLINT": ("integer", {"size": 16}),
-    "MEDIUMINT": ("integer", {"size": 32}),
-    "INTEGER": ("integer", {"size": 64}),
-    "REAL": ("float", {"size": 32}),
-    "FLOAT": ("float", {"size": 32}),
-    "DOUBLE": ("float", {"size": 64}),
-}
+
+def v2_to_gpkg_contents(dataset2):
+    """Generate a gpkg_contents meta item from a dataset v2"""
+    geom_columns = _get_geometry_columns(dataset2.schema)
+    is_spatial = bool(geom_columns)
+    result = {
+        "identifier": dataset2.get_meta_item("title"),
+        "description": dataset2.get_meta_item("description"),
+        "table_name": dataset2.tree.name,
+        "data_type": "features" if is_spatial else "attributes",
+    }
+    if is_spatial:
+        result["srs_id"] = srs_str_to_int(
+            geom_columns[0].extra_type_info["geometrySRS"]
+        )
+    return result
+
+
+def v2_to_gpkg_geometry_columns(dataset2):
+    """Generate a gpkg_geometry_columns meta item from a dataset v2"""
+    geom_columns = _get_geometry_columns(dataset2.schema)
+    if not geom_columns:
+        return None
+
+    geom_column = geom_columns[0]
+    type_name, *zm = geom_column.extra_type_info["geometryType"].split(" ", 1)
+    srs_id = srs_str_to_int(geom_column.extra_type_info["geometrySRS"])
+    zm = zm[0] if zm else ""
+    z = 1 if "Z" in zm else 0
+    m = 1 if "M" in zm else 0
+    return {
+        "table_name": dataset2.tree.name,
+        "column_name": geom_column.name,
+        "geometry_type_name": type_name,
+        "srs_id": srs_id,
+        "z": z,
+        "m": m,
+    }
+
+
+def v2_to_gpkg_spatial_ref_sys(dataset2):
+    """Generate a gpkg_spatial_ref_sys meta item from a dataset v2"""
+    geom_columns = _get_geometry_columns(dataset2.schema)
+    if not geom_columns:
+        return []
+
+    srs_str = geom_columns[0].extra_type_info["geometrySRS"]
+    srs_id = srs_str_to_int(srs_str)
+    definition = dataset2.get_meta_item(f"srs/{srs_str}.wkt")
+    # This should be more complicated too.
+    # TODO: srs_name, description.
+    return [
+        {
+            "srs_name": srs_str,  # This name is not quite right.
+            "definition": definition,
+            "organization": "EPSG",
+            "srs_id": srs_id,
+            "organization_coordsys_id": srs_id,
+        }
+    ]
+
+
+def v2_to_sqlite_table_info(dataset2):
+    return [_columnschema_to_gpkg(i, col) for i, col in enumerate(dataset2.schema)]
+
+
+def _get_geometry_columns(schema):
+    return [c for c in schema.columns if c.data_type == "geometry"]
+
+
+def srs_str_to_int(srs_str):
+    # This should be more complicated.
```

Comment:
what other forms could `srs_str` take here?

Feels like it should raise an error rather than 0 if it doesnt' understand the value given.

Also see `OGRImporter._get_meta_srid` which does some conversion via OGR's SRS objects to return an integer

Reply:
Well, right now the only SRS strings that this code deals with when creating gpkg from v2 datasets are those that were generated by the inverse function, srs_int_to_str, when creating the v2 dataset in the first place.
So right now this function is guaranteed to work. You're right, it should raise a ValueError if it can't handle it.

However, the int `srs_id` we get from geopackage during import / upgrade might not always be intended to be an EPSG identifier. The spec says that this could be `urn:ogc:def:crs:EPSG::2193` or basically any string eg `kx_custom_srs_512`, as long as a definition of the same name is included. But, it's not clear at the moment if or how we would end up with free form strings, since all our imports go via GPKG which requires an authority (string) + id (integer) format. Anyway, needs more work.

Reply:
-1/0 pretty commonly map onto "unknown". Which we shouldn't encourage, but kinda need to work.

Reply:
[`osgeo.osr.SpatialReference`](https://gdal.org/python/osgeo.osr.SpatialReference-class.html) is the way we validate/use spatial reference definitions. (`ImportFrom*` functions)